pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -0,0 +1,260 @@
+	}
+
+	// for each certbundle with generate==true make a CertificateRequest
+	for _, cb := range cd.Spec.CertificateBundles {
+		if cb.Generate == true {
+			domains := getDomainsForCertBundle(cb, cd)
+
+			certReq := createCertificateRequest(cb.Name, domains, cd)
+			desiredCRs = append(desiredCRs, certReq)
+		}
+	}
+
+	deleteCRs := []certmanv1alpha1.CertificateRequest{}
+
+	// find any extra certificateRequests and mark them for deletion
+	for i, currentCR := range currentCRs {
+		found := false
+		for _, desiredCR := range desiredCRs {
+			if desiredCR.Name == currentCR.Name {
+				found = true
+				break
+			}
+		}
+		if !found {
+			deleteCRs = append(deleteCRs, currentCRs[i])
+		}
+	}
+
+	// create/update the desired certificaterequests
+	for _, desiredCR := range desiredCRs {
+		currentCR := &certmanv1alpha1.CertificateRequest{}
+		searchKey := types.NamespacedName{Name: desiredCR.Name, Namespace: desiredCR.Namespace}
+
+		if err := r.client.Get(context.TODO(), searchKey, currentCR); err != nil {
+			if errors.IsNotFound(err) {
+				// create
+				if err := controllerutil.SetControllerReference(cd, &desiredCR, r.scheme); err != nil {
+					logger.Error(err, "error setting owner reference", "certrequest", desiredCR.Name)
+					return err
+				}
+
+				if err := r.client.Create(context.TODO(), &desiredCR); err != nil {
+					logger.Error(err, "error creating certificaterequest")
+					return err
+				}
+			} else {
+				logger.Error(err, "error checking for existing certificaterequest")
+				return err
+			}
+		} else {
+			// update or no updated needed
+			if !reflect.DeepEqual(currentCR.Spec, desiredCR.Spec) {
+				if err := r.client.Update(context.TODO(), currentCR); err != nil {
+					logger.Error(err, "error updating certificaterequest", "certrequest", currentCR.Name)
+					return err
+				}
+			} else {
+				logger.Info("no updated needed for certificaterequest", "certrequest", desiredCR.Name)
+			}
+		}
+	}
+
+	// delete the extra certificaterequests
+	for _, deleteCR := range deleteCRs {
+		if err := r.client.Delete(context.TODO(), &deleteCR); err != nil {
+			logger.Error(err, "error deleting CertificateRequest that is no longer needed", "certrequest", deleteCR.Name)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *ReconcileClusterDeployment) getCurrentCertificateRequests(cd *hivev1alpha1.ClusterDeployment, logger logr.Logger) ([]certmanv1alpha1.CertificateRequest, error) {
+	certReqsForCluster := []certmanv1alpha1.CertificateRequest{}
+
+	// get all CRs in the cluster's namespace
+	currentCRs := &certmanv1alpha1.CertificateRequestList{}
+	if err := r.client.List(context.TODO(), &client.ListOptions{Namespace: cd.Namespace}, currentCRs); err != nil {
+		logger.Error(err, "error listing current CertificateRequests")
+		return certReqsForCluster, err
+	}
+
+	// now filter out the ones that are owned by the cluster we're processing
+	for i, cr := range currentCRs.Items {
+		for _, owner := range cr.GetOwnerReferences() {
```

Comment:
You can use metav1.IsControlledBy()

Reply:
done. (also updated the mocking in the test cases to that metav1.IsControlledBy() works)